CRM/Upgrade/Incremental/php/FourSeven.php

```diff
@@ -478,6 +478,9 @@ public function upgrade_4_7_28($rev) {
           3 => $endId,
         ));
         $this->addTask($title, 'updateContributionInvoiceNumber', $startId, $endId, $invoicePrefix);
+        if (civicrm_api3('Setting', 'getvalue', array('name' => 'contact_undelete', 'group' => 'CiviCRM Preferences'))) {
+          $this->addTask('CRM-21439 Add assignee to deleted contact activity records if undelete is true', 'addAssigneetoDeletedContactActivity');
+        }
```

Comment:
Instead of adding the task once and leaving it to loop through "batches" - see example above for how to add the task once per batch. This ensures the website doesn't timeout when running the upgrade gui.